test/extended/two_node/tnf_node_replacement.go

```diff
@@ -0,0 +1,1508 @@
+	o.Expect(err).To(o.BeNil(), "Surviving node IP validation failed: %v", err)
+
+	klog.V(2).Infof("Target node %s IP: %s", targetNodeName, targetNodeIP)
+	klog.V(2).Infof("Surviving node %s IP: %s", survivingNodeName, survivingNodeIP)
+
+	return targetNodeIP, survivingNodeIP
+}
+
+// getNodeInternalIP gets the internal IP address of a node using JSON output for robust parsing
+func getNodeInternalIP(oc *exutil.CLI, nodeName string) (string, error) {
+	// Get node details in JSON format
+	nodeOutput, err := oc.AsAdmin().Run("get").Args("node", nodeName, "-o", "json").Output()
+	if err != nil {
+		return "", fmt.Errorf("failed to get node %s details: %v", nodeName, err)
+	}
+
+	// Parse the JSON into a Node struct
+	var node corev1.Node
+	if err := utils.UnmarshalJSON(nodeOutput, &node); err != nil {
+		return "", fmt.Errorf("failed to parse node JSON for %s: %v", nodeName, err)
+	}
+
+	// Find the InternalIP address from the node's status
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			if addr.Address == "" {
+				return "", fmt.Errorf("node %s has empty internal IP address", nodeName)
+			}
+			klog.V(4).Infof("Found internal IP %s for node %s", addr.Address, nodeName)
+			return addr.Address, nil
+		}
+	}
+
+	return "", fmt.Errorf("could not find internal IP address for node %s", nodeName)
+}
+
+// computeGatewayIP computes the gateway IP address from a node IP address.
+// This follows the same logic as libvirt bridges, where the gateway is the first
+// address in the subnet.
+//
+// For IPv4: 192.168.111.20 -> 192.168.111.1
+// For IPv6: fd00::20 -> fd00::1
+//
+// Parameters:
+//   - nodeIP: The IP address of a node (IPv4 or IPv6)
+//
+// Returns:
+//   - string: The gateway IP address
+//   - error: Any error that occurred during computation
+func computeGatewayIP(nodeIP string) (string, error) {
+	// Parse the IP address to validate and determine if it's IPv4 or IPv6
+	ip := net.ParseIP(nodeIP)
+	if ip == nil {
+		return "", fmt.Errorf("invalid IP address: %s", nodeIP)
+	}
+
+	// Check if it's IPv4 or IPv6
+	if ip.To4() != nil {
+		// IPv4: Set last byte to 1
+		// Example: 192.168.111.20 -> 192.168.111.1
+		ip = ip.To4()
+		ip[3] = 1
+		return ip.String(), nil
+	}
+
+	// IPv6: Set last two bytes to 1
+	// Example: fd00::20 -> fd00::1
+	// Example: fd00:0:0:0:0:0:0:20 -> fd00:0:0:0:0:0:0:1
+	ip[len(ip)-2] = 0
+	ip[len(ip)-1] = 1
+	return ip.String(), nil
+}
+
+// getNodeMACAddress retrieves the MAC address for a node from its BareMetalHost
+func getNodeMACAddress(oc *exutil.CLI, nodeName string) string {
+	// Find the BareMetalHost name using regex pattern matching
+	bmhName := findObjectByNamePattern(oc, bmhResourceType, machineAPINamespace, nodeName, "")
+
+	// Get the BareMetalHost YAML to extract the MAC address
+	bmhOutput, err := oc.AsAdmin().Run("get").Args(bmhResourceType, bmhName, "-n", machineAPINamespace, "-o", yamlOutputFormat).Output()
```

Comment:
Nice, now that we have the helper functions, we should be able to use the `GetBMH` function here